src/metabase/models/field.clj

```diff
@@ -119,6 +119,28 @@
     ;; otherwise we need to fetch additional info about Field's Table. This is cached for 5 seconds (see above)
     (perms-objects-set* table-id)))
 
+(defn- maybe-update-double [maybe-double-value]
+  (if (string? maybe-double-value)
+    (u/ignore-exceptions (Double/parseDouble maybe-double-value))
+    maybe-double-value))
+
+(defn- update-special-numeric-values
+  "When fingerprinting decimal columns, NaN and Infinity values are possible. Serializing these values to JSON just
+  yields a string, not a value double. This function will attempt to coerce any of those values to double objects"
+  [fingerprint]
+  (if (get-in fingerprint [:type :type/Number])
```

Comment:
Some time back I added `util/update-in-when`, which would clean up this a bit. Since it's not a super hot path, I'd be tempted to just thread those instead of nested updates.